Sources/FingerprintJSTests/Tests/CellularNetworkInfoHarvesterTests.swift

```diff
@@ -0,0 +1,125 @@
+import XCTest
+
+@testable import FingerprintJS
+
+#if os(iOS)
+final class CellularNetworkInfoHarvesterTests: XCTestCase {
+    private var cellularServiceInfoProviderSpy: CellularServiceInfoProvidingSpy!
+
+    private var sut: CellularNetworkInfoHarvester!
+
+    override func setUp() {
+        super.setUp()
+        cellularServiceInfoProviderSpy = .init()
+        sut = .init(
+            cellularServiceInfoProvider: cellularServiceInfoProviderSpy
+        )
+    }
+
+    override func tearDown() {
+        sut = nil
+        cellularServiceInfoProviderSpy = nil
+        super.tearDown()
+    }
+
+    func test_givenNoCelluarProviders_whenMobileCountryCodes_thenReturnsEmptyArray() {
+        // given
+        cellularServiceInfoProviderSpy.cellularProvidersReturnValue = []
+
+        // when
+        let mobileCountryCodes = sut.mobileCountryCodes
+
+        // then
+        XCTAssertTrue(mobileCountryCodes.isEmpty)
+        XCTAssertEqual(1, cellularServiceInfoProviderSpy.cellularProvidersCallCount)
+    }
+
+    func test_givenNoCelluarProviders_whenMobileNetworkCodes_thenReturnsEmptyArray() {
+        // given
+        cellularServiceInfoProviderSpy.cellularProvidersReturnValue = []
+
+        // when
+        let mobileNetworkCodes = sut.mobileNetworkCodes
+
+        // then
+        XCTAssertTrue(mobileNetworkCodes.isEmpty)
+        XCTAssertEqual(1, cellularServiceInfoProviderSpy.cellularProvidersCallCount)
+    }
+
+    func test_givenTwoCelluarProviders_whenMobileCountryCodes_thenReturnsArrayWithTwoElements() {
+        // given
+        let firstProvider = CarrierInfoProvidingSpy(mobileCountryCode: "001")
+        let secondProvider = CarrierInfoProvidingSpy(mobileCountryCode: "048")
+        cellularServiceInfoProviderSpy.cellularProvidersReturnValue = [
+            firstProvider,
+            secondProvider,
+        ]
+
+        // when
+        let mobileCountryCodes = sut.mobileCountryCodes
+
+        // then
+        XCTAssertEqual(["001", "048"], mobileCountryCodes)
+        XCTAssertEqual(1, cellularServiceInfoProviderSpy.cellularProvidersCallCount)
```

Comment:
Okay, let's break it down step-by-step.

> why should we care that the unit did call a specific external API and didn't call anything else?
By doing so we are also testing for the undesired side-effects, which is super important. Let's use `CellularNetworkInfoHarvester` and a call to `mobileCountryCodes` as an example. One could easily make a mistake when introducing changes and make `mobileCountryCodes` call `\.mobileNetworkCode` instead of `\.mobileCountryCode` on each `CarrierInfoProviding` instance, because they are similarly named and they return the value of the same type. With the current approach to writing unit tests, we will not only detect that `mobileCountryCodes` no longer returns the expected value, but the tests will also pinpoint the root cause of the issue, which is very useful. One could also introduce silly code that would needlessly call `\.mobileCountryCode` twice and, once again, the current unit tests would catch that. Obviously, such a comprehensive method of writing unit tests comes with its cost of adding a couple of additional assertions, but at least to me the confidence gained from working with well-tested code is worth it.

> we don't want to modify the test ideally when we changed the internal implementation but the API didn't change
And with the current assertions we most likely won't be modifying those tests, unless we introduce breaking changes to the API itself. The thing is that when SUT has certain dependency that provides it with some data to operate on, it is reasonable to make design decisions on how many times the given dependency's API shall be called in a certain test scenario. Once again, referring back to the previous example, would it make sense to call `\.mobileNetworkCode` on each `CarrierInfoProviding` instance when all we want is `mobileCountryCodes`? Most likely not, and in fact it seems so irrational that we can make an assertion claiming that it should never happen. This way the tests also guard us from making incorrect design decisions or at least ensure that we will think twice before writing the code from the given example.

> I believe there are specific cases where Spy usage makes sense and is beneficial...
Actually, when following the [Test Double](https://en.wikipedia.org/wiki/Test_double) approach, the Stubs and Spies are usually the two most frequently used types of objects when writing unit tests. Of course, we could simplify the naming and instead call every object a Mock, but it won't change the principles described in the previous two paragraphs.